src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -0,0 +1,140 @@
+    private readonly object _gate = new();
+
+    /// <summary>
+    /// An event raised whenever the current solution is changed.
+    /// </summary>
+    public event EventHandler<WorkspaceChangeEventArgs> WorkspaceChanged
+    {
+        add => AddEventHandler(value, WorkspaceEventType.WorkspaceChange);
+        remove => RemoveEventHandler(value, WorkspaceEventType.WorkspaceChange);
+    }
+
+    /// <summary>
+    /// An event raised whenever the workspace or part of its solution model
+    /// fails to access a file or other external resource.
+    /// </summary>
+    public event EventHandler<WorkspaceDiagnosticEventArgs> WorkspaceFailed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.WorkspaceFailed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.WorkspaceFailed);
+    }
+
+    /// <summary>
+    /// An event that is fired when a <see cref="Document"/> is opened in the editor.
+    /// </summary>
+    public event EventHandler<DocumentEventArgs> DocumentOpened
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentOpened);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentOpened);
+    }
+
+    /// <summary>
+    /// An event that is fired when any <see cref="TextDocument"/> is opened in the editor.
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentOpened
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+    }
+
+    /// <summary>
+    /// An event that is fired when a <see cref="Document"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<DocumentEventArgs> DocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when any <see cref="TextDocument"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when the active context document associated with a buffer 
+    /// changes.
+    /// </summary>
+    public event EventHandler<DocumentActiveContextChangedEventArgs> DocumentActiveContextChanged
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+    }
+
+    private void AddEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
```

Comment:
```suggestion
    private void AddLegacyEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
```

Or equivalent, since otherwise we have a bunch of similarly sounding methods.